### PR TITLE
Adding HA support for mgmt-gateway and oauth2-proxy services

### PR DESCRIPTION
**Pull request by rkachach**
## Motivation

The current architecture of `mgmt-gateway` introduced in previous PRs (https://github.com/ceph/ceph/pull/57535 and https://github.com/ceph/ceph/pull/58460) lacks fault tolerance due to the absence of high availability mechanisms, making critical components like nginx and oauth2-proxy daemons single points of failure. This results in the risk of the entire system becoming nonoperational if one of these components fails, which is a significant concern for critical production environments.

This PR includes changes for:

- Add support for `mgmt-gateway` and `oauth2-proxy` high availability setup
- Improvements to reduce monitoring daemons restarts due to mgr failover
- use the stable `md5_hash` instead of `hash` to avoid changes in deps after a mgr failover
- Provide an internal end-point for `service_discovery`
- Fix for issue: https://tracker.ceph.com/issues/68377
- More UT for mgmt-gateway and oauth2-proxy services

## Architecture changes

To address HA concerns, the proposed changes introduce redundancy for both mgmt-gateway and oauth2-proxy services, enhancing system resilience. The updated architecture deploys multiple mgmt-gateway instances in an active/standby clustered setup, ensuring seamless failover via keepalived. Additionally, oauth2-proxy is now served by multiple stateless instances, with each nginx load-balancing requests across these instances using round-robin. These changes aim to eliminate single points of failure, providing robust high availability across the entire system.

This PR also includes some additional changes to enhance overall system stability by minimizing daemon restarts due to `mgr` failover. Previously, daemons such as Prometheus needed to be reconfigured in the event of a mgr failover, but this is no longer necessary, as `service_discovery` is now running behind the internal nginx server, which automatically handles HA for it.

The following diagram depicts the new architecture:

image

Internal service HA mechanism will be still the same as depicted in the below diagram. However, instead of reaching directly the `mgmt-gateway` internal endpoint, services will be configured to use the virtual-ip provided in the spec instead i.e `https://<virtual_ip>:<internal_port>/internal/prometheus` . This way, mgmt-gateway HA mechanism is kept transparent for other services.

image

Following is an example of the spec files needed to deploy mgmt-gateway in HA setup. The ingress is configured in `keepalive_only`  mode and labels are used for placement to make sure any movement in the `mgmt-gateway` daemons is also replicated for the corresponding keepalived as well. The number of the deployed instances is determined by the number of hosts with the `mgmt` label on them.

## Testing

To test the new functionality in combination with authentication, the user has first to enable SSO/Oauth2 on the dashboard by using the following cmd:
```
ceph dashboard sso enable oauth2 
```

Once enabled, we can use a spec that includes all the needed components. Before proceding with the following configuration make sure you have previously configured your external IDP (keycloak i.e)  accordingly and gathered all the information needed to configure the `oauth2-proxy` service. Following is a complete spec that contains all the needed components:

```
---
service_type: host
hostname: ceph-node-1
addr: 192.168.100.101
labels:
  - mgmt
---
service_type: host
hostname: ceph-node-2
addr: 192.168.100.102
labels:
  - mgmt
---
service_type: mgmt-gateway
placement:
  label: mgmt
spec:
  enable_auth: true
  virtual_ip: 192.168.100.220
---
service_type: oauth2-proxy
placement:
  label: mgmt
spec:
  provider_display_name: "My OIDC Provider"
  client_id: "oauth2-proxy"
  client_secret: "vwCLTITb2rtem583523uXcuSNItDsz1I"
  oidc_issuer_url: "http://192.168.100.1:8080/realms/ceph"
  allowlist_domains:
    - 192.168.100.1:8080
---
service_type: ingress
service_id: ingress-mgmt-gw
placement:
  label: mgmt
virtual_ip: 192.168.100.220
backend_service: mgmt-gateway
keepalive_only: true
```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by rkachach** [comment]
jenkins retest this please

**Review by jmolmo — APPROVED** [review]
I was not able to find any issue in the code!!.. It looks nice ... Let's try it.

The explanation in the comments of this PR about how to configure HA for the mgmt gateway service is really good. 
Do not forget to copy it in the **documentation** ;-)

**Comment by rkachach** [comment]
jenkins retest this please

**Comment by rkachach** [comment]
> I was not able to find any issue in the code!!.. It looks nice ... Let's try it.
> 
> The explanation in the comments of this PR about how to configure HA for the mgmt gateway service is really good. Do not forget to copy it in the **documentation** ;-)

I added some docs for the new functionality :+1:

File: doc/cephadm/services/mgmt-gateway.rst
Comment (phlogistonjohn) [inline doc/cephadm/services/mgmt-gateway.rst:61]: `can also be` the also here disrupts the flow of the sentence and seems unnecessary.
Maybe: "... can be deployed using multiple stateless instances" works better?

**Review by phlogistonjohn — COMMENTED** [review]
I looked at the doc changes and only the doc changes, just to see if they matched my understanding of what has been discussed in various meetings. I have to suggest is a minor grammar improvement (IMO).

File: doc/cephadm/services/mgmt-gateway.rst
Comment (rkachach) [reply doc/cephadm/services/mgmt-gateway.rst:61]: I adjusted the sentence :+1:

File: src/pybind/mgr/cephadm/module.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/module.py:821]: @adk3798 these functions used to be part of the `mgmt_gateway.py` but since the introduction of the`oauth2-proxy` which needs to call the `get_mgmt_gw_external_endpoint` I moved them to the module.py to avoid circular dependencies between mgmt_gateway and oauth2_proxy services.

File: src/pybind/mgr/cephadm/module.py
Comment (rkachach) [reply src/pybind/mgr/cephadm/module.py:847]: @adk3798 the logic for these functions used to be separated but then I did some refactoring bcz they have a big common section. However, I still want to keep explicit differentiation (without relying on the parameter) between the two calls. In that sense I preferred having this small wrapper rather having to specify the `is_internal` in all the calls to these functions.

**Review by adk3798 — APPROVED** [review]

**Comment by adk3798** [comment]
jenkins test windows

**Review by anthonyeleven — APPROVED** [review]
Note to self: wordsmith after merge

**Comment by rkachach** [comment]
jenkins test dashboard cephadm

**Comment by rkachach** [comment]
jenkins test dashboard cephadm

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-10-22_07:55:35-orch:cephadm-wip-adk-testing-2024-10-21-1626-distro-default-smithi/

Failures:
- 2x smb stdb test SELinux denials / stray daemon warnings, known issue
- 3x mds_upgrade_sequence `in cluster log` / timeout issues, known issue
- 2x rgw-ingress timeout failures, known issue

Nothing to block merging

**Pull request merged**
